MobiusCore/Source/EffectHandlers/_EffectHandlerConnection.swift

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2019 Spotify AB.
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+public typealias SideEffect = () -> Void
+// swiftlint:disable type_name
+/// An `_EffectHandlerConnection` describes the lifecycle of an effect handler. It can receive effects, and output events to its `output` `Consumer`.
+/// The `_` in the name denotes that while this class needs to be exposed publicly, you should not depend on it directly as a user of Mobius.
+public final class _EffectHandlerConnection<Effect, Event>: Disposable {
+    private let canHandleFn: (Effect, @escaping Consumer<Event>) -> SideEffect?
+    // We cannot know if this `Consumer` is internally thread-safe. The thread-safety is therefore delegated to the
+    // `output` instance function.
+    private var unsafeOutput: Consumer<Event>?
+    // We cannot know if this `Disposable` is internally thread-safe. The thread-safety is therefore delegated to the
+    // `dispose` instance function.
+    private let unsafeDispose: Disposable
+    private let lock = Lock()
+
+    init(
+        canHandle canHandleFn: @escaping (Effect, @escaping Consumer<Event>) -> SideEffect?,
+        output unsafeOutput: @escaping Consumer<Event>,
+        disposable unsafeDispose: Disposable
+    ) {
+        self.canHandleFn = canHandleFn
+        self.unsafeOutput = unsafeOutput
+        self.unsafeDispose = unsafeDispose
+    }
+
+    /// Determine if a given effect can be handled.
+    /// If an effect can be handled, a `SideEffect` is returned. Calling this `SideEffect` will cause the effect to be performed
+    /// If the effect cannot be handled, nil is returned.
+    /// 
+    /// Note: Execution of the `SideEffect` function should never be deferred. Executing a `SideEffect` after `dispose`
+    /// has returned may cause runtime exceptions. `SideEffect` itself may internally be concurrent.
+    ///
+    /// - Parameter effect: the effect in question
+    public func canHandle(_ effect: Effect) -> SideEffect? {
```

Comment:
I have said this before, but never got a response, so repeating - to me, `selectHandler` is a better name for this method. One of the key things it does is, in particular for merged things (I'm staying away from the name 'effect handler'), is to figure out what code to run. The name `canHandle` misses that aspect of what the method does.

Reply:
I don't agree. `selectHandler` suggests that the current effect handler is being selected. It's not though. It is only potentially being selected, if it _can_ _Handle_ the effect.
The issue with `canHandle` is that it suggests that a boolean is being returned. I think the type system mostly compensates for that lack of clarity.

I would like to see a better name for this, but I don't think `selectHandler` is it.

Reply:
Yes - I guess to me, the thing that's selecting a handler isn't a handler. :) But let's talk about it face to face, there's a better name somewhere.

Reply:
wouldn't `selectSideEffect` be the proper way to name it? afaik that's how `select` usually is used in other reactive frameworks eg. reflex or RxJs

Reply:
I'm not finding any evidence of that. Please provide links.

Reply:
@togi and I discussed this offline and settled on `sideEffectFor(_ effect: Effect) -> SideEffect?`

Reply:
- reflex has EventSelector https://hackage.haskell.org/package/reflex-0.2/candidate/docs/Reflex-Class.html that picks an event

- RxJs uses select as an alternative name for `map` https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/select.md

- to some degree  `select(2)` for unix file descriptors http://man7.org/linux/man-pages/man2/select.2.html isn't completely unrelated, but it does have some extra semantics

i don't have a really strong opinion on specific name, I mostly meant there are prior art to use `select` to describe a mapping a bit like this, but that if you use that name it probably would be referring to the thing you select (in this case a SideEffect) rather than what you select it from (a handler).

Either way I think it'd be good to pick a name that isn't using `can` unless it's returning a boolean.